Refactor foreach unary ops tests to use OpInfo

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 [WIP] Add type promotion to a fast path for foreach APIs
* #52448 [WIP] Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* **#49712 [wip] Refactor foreach unary ops tests to use OpInfo**

Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

Commit:
Refactor Unary Ops tests

[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -153,54 +120,881 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
-                # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
-                control_dtype = torch.float32 if (self.device_type == 'cuda' and
-                                                  (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_neg(self, device, dtype):
+        for N in N_values:
+            tensors = self._get_test_data(device, dtype, N)
```
We should really consider having OpInfo based tests because I think they will probably eliminate the need for a lot of this logic. I'm not sure if OpInfo covers all the operators that you have here, or how difficult it is to add a new OpInfo into the op database.

If we want to go ahead with the testing without using OpInfo, then I prefer the old testing more than the new testing. The older testing is a lot more concise, but has the following downsides:
1. the branches are a little difficult to follow and end up special casing certain operators. I.e., there isn't a clean "mechanism vs policy" split.
2. Everything runs in the same test case so it's hard to tell which operator is erroring out.

We can fix both of the above problems. For problem 1, I think we just need to add more to the unary_ops table:
```
 foreach_op, foreach_op_, torch_op, cpu_half, cuda_half, cpu_bf16, cuda_bf16, complex64/128, bool
```
this would separate the mechanism (the testing logic) from the policy (the operator table).

For the second problem, we could autogenerate the individual tests by iterating through the table and assigning to the test class. I am not sure how well this works with all the testing decorators though.

izdeby committed:
Update on "Refactor Unary Ops tests"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

[ghstack-poisoned]

Commit:
Update on "Refactor Unary Ops tests"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

[ghstack-poisoned]

Commit:
Update on "Refactor Unary Ops tests"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

[ghstack-poisoned]

Commit:
Update on "Refactor Unary Ops tests"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

[ghstack-poisoned]

Commit:
Update on "Refactor Unary Ops tests"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -944,6 +944,41 @@ def __init__(self,
+                 method,
+                 inplace,
+                 ref,  # torch reference function
+                 dtypes=floating_and_complex_types(),
+                 dtypesIfCPU=all_types_and_complex(),
+                 dtypesIfCUDA=floating_and_complex_types_and(torch.half),
```
Just wondering:
- Why are the dtypesIfCPU different from dtypesIfCUDA? Generally if something works on CPU I'd expect it to work on CUDA as well
- What does "dtypes" do? We've already specified what the dtypes for CPU and CUDA are

mruberry replied:
> Just wondering:
> 
> * Why are the dtypesIfCPU different from dtypesIfCUDA? Generally if something works on CPU I'd expect it to work on CUDA as well

Unfortunately that's often not the case for half and bfloat16 suppor.

> * What does "dtypes" do? We've already specified what the dtypes for CPU and CUDA are

We also run meta tests and XLA (TPU) tests and support other hardware using our tests, and dtypes is for them.

zou3519 replied:
> Unfortunately that's often not the case for half and bfloat16 support

Those exceptions make sense. I'm wondering if there is a reason why dtypesIfCPU includes integral types by default but dtypesIfCUDA doesn't include those by default

izdeby replied:
The problem with integral types and CUDA is that foreach doesn't allow type promotion right now (coming in next PRs of this stack). So with int input we cant return float result, therefore we have some restrictions that will be lifted later in this stack.

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -221,6 +224,9 @@ std::vector<Tensor> foreach_tensor_neg_cuda(TensorList tensors) {
     check_foreach_api_restrictions(tensors);
+    TORCH_CHECK(tensors[0].scalar_type() != kBool,
+                "Negation, the `-` operator, on a bool tensor is not supported. "
+                "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
```
Does `~` or `logical_not` work on TensorList right now?

zou3519 replied:
Should we improve the error message? Some things I am noticing:
1. If the user gets this error message, they did not actually use the `-` builtin operator. E.g. a user didn't do `tensors = [...]; other_tensors = -tensors`. So the "the `-` operator" part of the error message is confusing
2. The error message assumes that the user applied operations to a bool tensor. Maybe we should mention that the API being called is "_foreach_neg"
3. The error message shows up after a user calls `_foreach_neg`. But the error message suggests workarounds for a single tensor. This may be misunderstood to "just use ~ on your tensorlist"

Adding all the top together, maybe something like the following would make more sense?
```
_foreach_neg: There is a bool tensor in the passed-in TensorList.
Negation on a bool tensor is not supported. If you are trying to invert a mask, please
use the `~` or `logical_not()` operator on the individual tensors instead
```

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -944,6 +944,41 @@ def __init__(self,
                                             **kwargs)
         self.ref = ref
 
+def sample_inputs_foreach(self, device, dtype, N):
+    if dtype in [torch.bfloat16, torch.bool, torch.float16]:
+        tensors = [torch.randn((N, N), device=device).to(dtype) for _ in range(N)]
+    elif dtype in torch.testing.get_all_int_dtypes():
+        tensors = [torch.randint(1, 100, (N, N), device=device, dtype=dtype) for _ in range(N)]
+    else:
+        tensors = [torch.randn((N, N), device=device, dtype=dtype) for _ in range(N)]
+
+    return tensors
```
In the future it might be worth testing non-contiguous inputs. But this looks fine for now since it's a refactor of the existing foreach tests

mruberry replied:
Prefer using the test suite's make_tensor() to randint() or randn(), it will handle all dtypes for you. It also has a "discontiguous" kwarg for the future. (That arg will be correctly renamed to noncontiguous in the future, too.)

zou3519 commented on test/test_foreach.py:
```diff
@@ -157,90 +126,27 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
+
+            method = op.get_method()
+            inplace = op.get_inplace()
+            fe_res = method(tensors)
```
nit: What does `fe` in `fe_res` stand for? Maybe just `expected_res` would be clearer

zou3519 reviewed (commented):
Some questions but this looks like a reasonable refactor

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

zou3519 reviewed (commented):
The testing refactor LGTM. @mruberry do you have other comments about how OpInfo is being used?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1470,6 +1499,188 @@ def _make_tensor_helper(shape, low=None, high=None):
+                         safe_casts_outputs=False),
+
+    ForeachUnaryFuncInfo('_foreach_sqrt',
+                         method=torch._foreach_sqrt,
```
Definitely not a blocker for this PR, but one thing we do with the base OpInfos is attempt to programmatically acquire the method and inplace variants of an op, which this could do, too, if they weren't specified.

izdeby replied:
done

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1470,6 +1499,188 @@ def _make_tensor_helper(shape, low=None, high=None):
+                         dtypesIfCPU=floating_and_complex_types_and(torch.bfloat16),
+                         dtypesIfCUDA=floating_and_complex_types_and(torch.half)),
+
+    ForeachUnaryFuncInfo('_foreach_exp',
```
These names could probably also be auto-generated by default from a string like 'exp', and the reference op also acquired from the same string automatically.

izdeby replied:
Done

mruberry commented:
> The testing refactor LGTM. @mruberry do you have other comments about how OpInfo is being used?

This is a smart use of the OpInfo pattern. I made a couple suggestions for ideas that might be nice to have in the future, but this LGTM!

zou3519 reviewed (approved):
approving, but please see nit on the error message

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby committed:
Update on "[wip] Refactor foreach unary ops tests to use OpInfo"


Differential Revision: [D25673712](https://our.internmc.facebook.com/intern/diff/D25673712)

-----
- Updated foreach unary ops tests to use OpInfo

[ghstack-poisoned]

izdeby commented:
> approving, but please see nit on the error message

Done!
